File: android_review/app/src/main/java/com/example/myapplication/bottomnavigation/ui/home/IdleGoodsDetailInfoActivity.java
package com.example.myapplication.bottomnavigation.ui.home;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.LinearLayout;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myapplication.R;
import com.example.myapplication.adapter.CommentAdapter;
import com.example.myapplication.bean.CommentItem;
import com.example.myapplication.myview.MyTitleBar;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.UUID;


public class IdleGoodsDetailInfoActivity extends AppCompatActivity {
    private MyTitleBar idleGoodsDetailInfoMyTitleBar;
    private EditText editComment;
    private LinearLayout ll_leave_comment;
    private RecyclerView rv_comments;

    private List<CommentItem> initTestComments() {
        List<CommentItem> commentItems = new ArrayList<>();
        for (int i = 0; i < 10; i++) {
            CommentItem item = new CommentItem(UUID.randomUUID().toString());
            item.setContent("fake content" + i);
            item.setUsername("fake username" + i);
            item.setPortrait("fake content" + i);
            item.setDate(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date()));
            commentItems.add(item);
        }
        return commentItems;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_idle_goods_detail_info);

        idleGoodsDetailInfoMyTitleBar = findViewById(R.id.myTitleBar_idleGoodsDetailInfo);
        editComment = findViewById(R.id.et_commemt);
        ll_leave_comment = findViewById(R.id.ll_leave_comment);
        rv_comments = findViewById(R.id.rv_comments);
        rv_comments.setLayoutManager(new LinearLayoutManager(this));
        rv_comments.setAdapter(new CommentAdapter(this, initTestComments()));

        getSupportActionBar().hide();
        editComment.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                Log.d("onEditorAction", "" + v.getText());
            }
            return false;
        });
        ll_leave_comment.setOnClickListener(v -> {
            editComment.requestFocus();
            InputMethodManager inputManager = (InputMethodManager) editComment.getContext()
                    .getSystemService(Context.INPUT_METHOD_SERVICE);
            if (inputManager != null) {
                inputManager.showSoftInput(editComment, 0);
            }
        });

        // 更改顶部菜单栏标题
        idleGoodsDetailInfoMyTitleBar.setTvTitleText("闲置物详情");
        idleGoodsDetailInfoMyTitleBar.getTvForward().setVisibility(View.GONE);


    }
}

File: android_review/app/src/main/java/com/example/myapplication/activity/IdleGoodsDetailInfoActivity.kt
package com.example.myapplication.activity

import android.os.Bundle
import android.util.Log
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import cn.bmob.v3.BmobQuery
import cn.bmob.v3.exception.BmobException
import cn.bmob.v3.listener.QueryListener
import cn.bmob.v3.listener.UpdateListener
import com.afollestad.materialdialogs.DialogAction
import com.afollestad.materialdialogs.MaterialDialog
import com.example.myapplication.R
import com.example.myapplication.adapter.CommentAdapter
import com.example.myapplication.bean.Stuff
import com.example.myapplication.utils.Config
import kotlinx.android.synthetic.main.activity_idle_goods_detail_info.*
import java.util.*

class IdleGoodsDetailInfoActivity : AppCompatActivity() {
    private var stuff: Stuff? = null
    private var isCollected: Boolean = false


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_idle_goods_detail_info)
        supportActionBar?.hide()
        // 更改顶部菜单栏标题
        myTitleBar_idleGoodsDetailInfo.setTvTitleText("闲置物详情")
        myTitleBar_idleGoodsDetailInfo.tvForward?.visibility = View.GONE
        val bmobQuery = BmobQuery<Stuff>()
        bmobQuery.getObject("", object : QueryListener<Stuff>() {
            override fun done(p0: Stuff?, p1: BmobException?) {
                stuff = p0
                rv_comments.layoutManager = LinearLayoutManager(this@IdleGoodsDetailInfoActivity)
                rv_comments.adapter =
                    stuff?.comments?.let { CommentAdapter(this@IdleGoodsDetailInfoActivity, it) }
            }

        })

        et_commemt.setOnEditorActionListener { v: TextView, actionId: Int, event: KeyEvent? ->
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                Log.d("onEditorAction", "" + v.text)
            }
            false
        }
        ll_leave_comment?.setOnClickListener { v: View? ->
            et_commemt.requestFocus()
            val inputManager =
                et_commemt.context?.getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            inputManager.showSoftInput(et_commemt, 0)
        }
        tv_want.setOnClickListener {
            banWantDialog()
        }
        ll_collect.setOnClickListener {
            collectStuff()
        }

    }

    private fun collectStuff() {
        if (stuff == null) {
            Toast.makeText(this, "收藏失败，未获取商品详情", Toast.LENGTH_SHORT).show()
            return
        }
        if (!isCollected) {
            Config.user?.apply {
                collections?.add(stuff!!)
                update(object : UpdateListener() {
                    override fun done(p0: BmobException?) {
                        if (p0 != null) {
                            Log.w("collectStuff", "error  ${p0.message}")
                            return
                        }
                        Toast.makeText(
                            this@IdleGoodsDetailInfoActivity, "收藏成功",
                            Toast.LENGTH_SHORT
                        ).show()
                        iv_collect.setImageResource(R.drawable.ic_baseline_star_24)
                        isCollected = true
                    }
                })
            }
        } else {
            Config.user?.apply {
                collections?.remove(stuff)
                update(object : UpdateListener() {
                    override fun done(p0: BmobException?) {
                        if (p0 != null) {
                            Log.w("collectStuff", "error  ${p0.message}")
                            return
                        }
                        Toast.makeText(
                            this@IdleGoodsDetailInfoActivity, "取消收藏",
                            Toast.LENGTH_SHORT
                        ).show()
                        iv_collect.setImageResource(R.drawable.ic_baseline_star_border_24)
                        isCollected = false
                    }
                })
            }
        }

    }

    private fun banWantDialog() {
        MaterialDialog.Builder(this)
            .content(R.string.ban_want_hint)
            .positiveText(android.R.string.ok)
            .negativeText(android.R.string.cancel)
            .onPositive { _: MaterialDialog, _: DialogAction? ->
                Toast.makeText(this, "加入购物车", Toast.LENGTH_SHORT).show()
            }.show()
    }
}